contracts/src/groth16.rs

```diff
@@ -12,9 +12,82 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use alloy::sol_types::SolValue;
+use alloy::{hex, primitives::Bytes, sol_types::SolValue};
 use anyhow::Result;
-use risc0_zkvm::{sha::Digestible, Groth16ReceiptVerifierParameters};
+use risc0_zkvm::{
+    sha::Digestible, FakeReceipt, Groth16Receipt, Groth16ReceiptVerifierParameters, InnerReceipt,
+    MaybePruned, Receipt, ReceiptClaim,
+};
+
+alloy::sol!(
+    #![sol(all_derives)]
+    struct Seal {
+        uint256[2] a;
+        uint256[2][2] b;
+        uint256[2] c;
+    }
+);
+
+impl Seal {
+    fn flatten(self) -> Vec<u8> {
+        self.a
+            .iter()
+            .map(|x| x.to_be_bytes_vec())
+            .chain(
+                self.b
+                    .iter()
+                    .flat_map(|x| x.iter().map(|y| y.to_be_bytes_vec())),
+            )
+            .chain(self.c.iter().map(|x| x.to_be_bytes_vec()))
+            .flatten()
+            .collect()
+    }
+
+    /// Convert the [Seal] into a [Receipt] constructed with the given [ReceiptClaim] and
+    /// journal. The verifier parameters are optional and default to the current zkVM version.
+    pub fn to_receipt(
+        self,
+        claim: ReceiptClaim,
+        journal: impl AsRef<[u8]>,
+        verifier_parameters: Option<Groth16ReceiptVerifierParameters>,
+    ) -> Receipt {
+        let inner = risc0_zkvm::InnerReceipt::Groth16(Groth16Receipt::new(
+            self.flatten(),
+            MaybePruned::Value(claim),
+            verifier_parameters.unwrap_or_default().digest(),
+        ));
+        Receipt::new(inner, journal.as_ref().to_vec())
+    }
+}
+
+/// Decode a seal with selector as [Bytes] into a [Receipt] constructed with the given [ReceiptClaim]
+/// and journal. The verifier parameters are optional and default to the current zkVM version.
+pub fn decode_seal(
+    seal: Bytes,
+    claim: ReceiptClaim,
+    journal: impl AsRef<[u8]>,
+    verifier_parameters: Option<Groth16ReceiptVerifierParameters>,
+) -> Result<Receipt> {
+    let seal_bytes = seal.to_vec();
+    let (selector, stripped_seal) = seal_bytes.split_at(4);
+    // Fake receipt seal is 32 bytes
+    let receipt = if stripped_seal.len() == 32 {
+        if selector != [0u8; 4] {
+            return Err(anyhow::anyhow!(
+                "Invalid selector {} for fake receipt",
+                hex::encode(selector)
+            ));
+        };
+        Receipt::new(
+            InnerReceipt::Fake(FakeReceipt::new(claim)),
+            journal.as_ref().to_vec(),
+        )
+    } else {
```

Comment:
I don't think this function should handle fake receipts. It's currently an odd compromise, where it handles Groth16 and Fake receipts, but not set inclusion receipts and it would not handle e.g. Plonk receipts if we add them later. We probably needs two types of decoding. One would be a method that decodes to a specific type (e.g. Groth16, Fake, or SetInclusion), and another that decodes any seal in a list of known types. For the later, we will need a table included in the library mapping selectors to verifier parameter (digests). At the very least, that table should include Groth16, SetInclusion and Fake verifier parameters for the current version. We could additionally include previous versions in the this table as well. We've discussed this elsewhere, and it seems likely to be useful.